Add new columns for past country residence

quazi-h opened the pull request:
## Why
DAP has 10 fields to get the alternate past residence country for a participant.
They initially were allowing users to enter their own text strings and have since deprecated that field and are now providing a dropdown to choose a country form. In our old logic we were checking both. DAP has requested that we add 10 columns for the "text entry" past residences.
[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1484)

## This PR
- Added 10 new fields to the ResidentialEnvironmentTransformations
- Modified the parseCountry to beck for both the  "_01" vs "_01_only" RC variables
- Updated unit tests

quazi-h committed:
arh commit

aherbst-broad committed:
Fix

aherbst-broad committed:
Update tests

quazi-broad commented:
Two things:

- Drew made some updates on his branch that need to be pushed.
- The logic in transformations could be refined - may do this Monday

Somewhat irrelevant to the PR but won't adding 10 new columns for data that only a handful of records have cost more than its worth? (not complaining - just thinking that is a little overkill)

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1484-add-more-residence-cols

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/dog/ResidentialEnvironmentTransformations.scala:
```diff
@@ -28,12 +28,12 @@ object ResidentialEnvironmentTransformations {
+    only1Add: String
   ): Option[String] = {
-    rawRecord.getOptional(countryField).orElse(rawRecord.getOptional(countryFieldDropdown))
+    rawRecord.getOptional(multipleAdd).orElse(rawRecord.getOptional(only1Add))
```
it's a little unclear to me how the function name and argument names relate. specifically, a function that parses both fields but is named `parseOnlyCountry` is confusing to me.

could we either change these names or add comments clarifying what each argument of the function is?

quazi-broad replied:
Added comments to the function to explain the new args

ThePletch commented on hles/transformation/src/main/scala/org/broadinstitute/monster/dap/dog/ResidentialEnvironmentTransformations.scala:
```diff
@@ -58,38 +58,54 @@ object ResidentialEnvironmentTransformations {
+                .orElse(rawRecord.getOptional("de_country_01_only_dd"))
             else None,
+          dePastResidenceCountry2 =
+            if (pastCountryCount > 2) rawRecord.getOptional("de_country_02_dd") else None,
```
rather than copy-paste this line eighteen times, let's make a function for it and use that in all eighteen places instead.

quazi-broad replied:
I renamed the function and moved the if (count check) and else None into the function.
I feel like this could be even further simplified if we just iterated over a range [1-10] to check for a certain prefix + suffix of RC variable for a given column in our table, something like:
`dePastResidenceCountry[N]        = getPastResidenceCountry(prefix="de_country_",suffix="_dd",range=2-10)
dePastResidenceCountry[N]Text = getPastResidenceCountry(prefix="de_country_",suffix=None,range=2-10)`
but I think what we have now may be better because its easier to read and understand the RC variable names we are checking. Let me know your thoughts

ThePletch replied:
i think abstracting over the prefix/suffix would be overkill, especially since we'd need to manually assign each one to its specific attribute anyway.

ThePletch reviewed (changes requested):
couple of code cleanup suggestions to improve clarity and reduce repetition.

quazi-h committed:
Moved count checking logic and else None branching logic to the function we are using.
Renamed the function to be more descriptive and added some comments to explain arguments/
Added notes on why we are breaking convention when checking the first country.
Applied to past residence country fields, both the drop down and original text columns.
